// @adiwajshing/baileys/lib/Types/index.js
"use strict";

// Import modules
const { Auth } = require("./Auth");
const { GroupMetadata } = require("./GroupMetadata");
const { Chat } = require("./Chat");
const { Contact } = require("./Contact");
const { State } = require("./State");
const { Message } = require("./Message");
const { Socket } = require("./Socket");
const { Events } = require("./Events");
const { Product } = require("./Product");
const { Call } = require("./Call");
const { Signal } = require("./Signal");

// Import and re-export types
export type { Auth as AuthType, GroupMetadata as GroupMetadataType, Chat as ChatType, Contact as ContactType, State as StateType, Message as MessageType, Socket as SocketType, Events as EventsType, Product as ProductType, Call as CallType, Signal as SignalType };

// Define DisconnectReason enum
export const enum DisconnectReason {
    connectionClosed = 428,
    connectionLost = 408,
    connectionReplaced = 440,
    timedOut = 408,
    loggedOut = 401,
    badSession = 500,
    restartRequired = 515,
    multideviceMismatch = 411,
}
