// @jimp/plugin-shadow/babel.config.js
import type { API } from 'babel-types';

module.exports = (api: API) => {
  api.cache(true);

  return {
    presets: [
      [
        '@babel/env',
        {
          useBuiltIns: 'usage',
          corejs: 3,
          targets: {
            node: 'current',
            browsers: ['last 2 versions', 'ie >= 11'],
          },
        },
      ],
    ],

    plugins: [
      '@babel/proposal-class-properties',
      '@babel/syntax-object-rest-spread',
      process.env.BABEL_ENV !== 'module' && 'add-module-exports',
      [
        'transform-inline-environment-variables',
        {
          include: ['BABEL_ENV', 'ENV'],
          exclude: ['NO_Transform'],
        },
      ],
    ].filter(Boolean),

    env: {
      test: {
        plugins: ['istanbul', { forceAllTransforms: true }],
      },
      development: {
        plugins: [
          process.env.ENV !== 'browser' && 'source-map-support',
          process.env.ENV !== 'browser' && { debug: true },
        ].filter(Boolean),
      },
      module: {
        presets: [['@babel/env', { modules: false }]],
      },
    },
  };
};
